_posts/2022-09-28-import-big-csv.md

```diff
@@ -0,0 +1,306 @@
+into Parquet using the Single file API, you would use:
+
+```r
+write_parquet(data, "~/dataset/data.parquet")
+```
+
+Creating this file takes about 85 seconds on my system. The resulting
+file is about 9.5 GB. You can then load it again the next time you
+need to work with it usiing:
+
+```r
+data <- read_parquet("~/dataset/data.parquet", as_data_frame = FALSE)
+```
+
+Reading this file takes about 25 seconds.
+
+A query looking at the number of occurences for each of the 10
+distinct values in a column takes half a second on the dataset loaded
+in memory (it is about the same timing whether the data is loaded from
+`read_csv_arrow()` or from `read_parquet()` because the in-memory
+representation is the same).
+
+Let's now use the Dataset API.
+
+## The Dataset API in R
+
+We will read the large CSV file with `open_dataset()`. This function can
+be pointed to a folder with lots of files but it can also read a
+single file.
+
+```r
+data <- open_dataset("~/dataset/path_to_file.csv")
+```
+
+With my 15GB file, it takes 0.05 seconds to "read" the file. It is
+fast because the data does not get loaded in memory. `open_dataset()`
+scans the content of the file, figures out the name of the columns and
+their data types.
+
+So running the same query as above, counting the number of unique
+values in a column, takes 18 seconds. It is slow because to perform
+this query, the query engine needs to actually read the data. It is
+the same result that we had found in a previous post. Running queries
+directly on a CSV file is slow. In that post, we had also found that
+storing the data in the parquet format sped things up. Let's now
+convert this dataset into the Parquet format using the Dataset API.
+
+Partitioning a Parquet dataset could help with query performance. From
+the example using the Single file API, we saw that we did not get
+great performance with a single Parquet file. The particular dataset I
+have on hands does not have any obvious variable we can use to
+partition the data. Instead, we can use the `max_rows_per_file`
+argument to limit how large each parquet file is. I found that (at
+least with the dataset I'm currently working with), limiting the
+number of rows to 10 million per file seemed like a good
+compromise. Each file is about 720 MB which is close to the file sizes
+in the NYC taxi dataset which is partitioned by year and month.
+
+```r
+write_dataset(
+  data,
+  format = "parquet",
+  path = "~/datasets/my-data/",
+  max_rows_per_file = 1e7
+)
+```
+
+Writing these files on my system takes about 50 seconds.
+
+Next time we want to analyze this data, we can load these files with:
+
+```r
+data <- open_dataset(
+  "~/datasets/my-data"
+  )
+```
+
+It takes about the same amount of time as scanning the CSV files (0.02
+seconds).
+
+Counting the number of occurences for each of the 10 unique values in
+a column takes just 1 second. One second to summarize 140 million
```

Comment:
Could you use bold on the "just 1 second," then directly add together the read+query time for both to really just hammer home the comparison? As is, I have to scroll up to remember the read+query cost for the single-file API.